File: src/cli/scan_command.zig
pub const ScanCommand = struct {
    pub fn exec(ctx: Command.Context) !void {
        const cli = try PackageManager.CommandLineArguments.parse(ctx.allocator, .scan);

        const manager, const cwd = PackageManager.init(ctx, cli, .scan) catch |err| {
            if (err == error.MissingPackageJSON) {
                Output.errGeneric("No package.json found. 'bun pm scan' requires a lockfile to analyze dependencies.", .{});
                Output.note("Run \"bun install\" first to generate a lockfile", .{});
                Global.exit(1);
            }
            return err;
        };
        defer ctx.allocator.free(cwd);

        try execWithManager(ctx, manager, cwd);
    }

    pub fn execWithManager(ctx: Command.Context, manager: *PackageManager, original_cwd: []const u8) !void {
        if (manager.options.security_scanner == null) {
            Output.prettyErrorln("<r><red>error<r>: no security scanner configured", .{});
            Output.pretty(
                \\
                \\To use 'bun pm scan', configure a security scanner in bunfig.toml:
                \\  [install.security]
                \\  scanner = "<cyan>package_name<r>"
                \\
                \\Security scanners can be npm packages that export a scanner object.
                \\
            , .{});
            Global.exit(1);
        }

        Output.prettyError(comptime Output.prettyFmt("<r><b>bun pm scan <r><d>v" ++ Global.package_json_version_with_sha ++ "<r>\n", true), .{});
        Output.flush();

        const load_lockfile = manager.lockfile.loadFromCwd(manager, ctx.allocator, ctx.log, true);
        if (load_lockfile == .not_found) {
            Output.errGeneric("Lockfile not found. Run 'bun install' first to generate a lockfile.", .{});
            Global.exit(1);
        }
        if (load_lockfile == .err) {
            Output.errGeneric("Error loading lockfile: {s}", .{@errorName(load_lockfile.err.value)});
            Global.exit(1);
        }

        const security_scan_results = security_scanner.performSecurityScanForAll(manager, ctx, original_cwd) catch |err| {
            Output.errGeneric("Could not perform security scan (<d>{s}<r>)", .{@errorName(err)});
            Global.exit(1);
        };

        if (security_scan_results) |results| {
            defer {
                var results_mut = results;
                results_mut.deinit();
            }

            security_scanner.printSecurityAdvisories(manager, &results);

            if (results.hasAdvisories()) {
                Global.exit(1);
            } else {
                Output.pretty("<green>No advisories found<r>\n", .{});
            }
        }

        Global.exit(0);
    }
};

const security_scanner = @import("../install/PackageManager/security_scanner.zig");
const Command = @import("../cli.zig").Command;
const PackageManager = @import("../install/install.zig").PackageManager;

const bun = @import("bun");
const Global = bun.Global;
const Output = bun.Output;

File: src/runtime/cli/scan_command.rs
use crate::Command;
use crate::cli::package_manager_command::PackageManagerCommand;
use bun_core::{Global, Output};
use bun_install::package_manager::security_scanner;
use bun_install::{Lockfile, PackageManager};

pub(crate) struct ScanCommand;

impl ScanCommand {
    pub(crate) fn exec_with_manager(
        ctx: Command::Context,
        manager: &mut PackageManager,
        original_cwd: &[u8],
    ) -> crate::Result<()> {
        if manager.options.security_scanner.is_none() {
            bun_core::pretty_errorln!("<r><red>error<r>: no security scanner configured");
            bun_core::pretty!(
                "\n\
                 To use 'bun pm scan', configure a security scanner in bunfig.toml:\n  \
                 [install.security]\n  \
                 scanner = \"<cyan>package_name<r>\"\n\
                 \n\
                 Security scanners can be npm packages that export a scanner object.\n"
            );
            Global::exit(1);
        }

        bun_core::pretty_error!(
            "<r><b>bun pm scan <r><d>v{}<r>\n",
            Global::package_json_version_with_sha,
        );
        Output::flush();

        // Reshaped for borrowck — `manager.lockfile.load_from_cwd(&mut self,
        // Some(manager), log)` would alias `&mut *manager.lockfile` with `&mut *manager`.
        // Project disjoint raw pointers from the singleton first; `load_from_cwd` only
        // reads `manager.options`/migration helpers and never re-borrows `manager.lockfile`.
        {
            let log_level = manager.options.log_level;
            let pm_ptr: *mut PackageManager = manager;
            // SAFETY: `manager.log` is set non-null by `PackageManager::init`.
            let log: &mut bun_ast::Log = unsafe { &mut *(*pm_ptr).log };
            // SAFETY: `lockfile` is the owned `Box<Lockfile>` field on the singleton;
            // no other live `&mut Lockfile` exists at this point.
            let lockfile: &mut Lockfile = unsafe { &mut *(*pm_ptr).lockfile };
            let load_result = lockfile.load_from_cwd(
                // SAFETY: see comment above — `load_from_cwd` accesses `manager`
                // fields disjoint from `lockfile`.
                Some(unsafe { &mut *pm_ptr }),
                log,
                true,
            );
            PackageManagerCommand::handle_load_lockfile_errors_for(&load_result, log_level, "scan");
        }

        let security_scan_results =
            match security_scanner::perform_security_scan_for_all(manager, &mut *ctx, original_cwd)
            {
                Ok(v) => v,
                Err(e) => {
                    Output::err_generic("Could not perform security scan (<d>{s}<r>)", (e.name(),));
                    Global::exit(1);
                }
            };

        if let Some(results) = security_scan_results {
            // `defer { var results_mut = results; results_mut.deinit(); }` — Drop handles it.

            security_scanner::print_security_advisories(manager, &results);

            if results.has_advisories() {
                Global::exit(1);
            } else {
                bun_core::pretty!("<green>No advisories found<r>\n");
            }
        }

        Global::exit(0);
    }
}
